From 92d035cd7b459011bb085c0df620032ecc092099 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 26 May 2013 10:24:28 -0400
Subject: [PATCH] Entity Spawn Events

These events fire any time an entity is added/removed to the world, so we do not
have to check multiple locations for entities being added/removed from the world.
---
 .../empireminecraft/customevents/EntityEvents.java | 53 ++++++++++++++++++++++
 src/main/java/net/minecraft/server/World.java      |  2 +
 2 files changed, 55 insertions(+)
 create mode 100644 src/main/java/com/empireminecraft/customevents/EntityEvents.java

diff --git a/src/main/java/com/empireminecraft/customevents/EntityEvents.java b/src/main/java/com/empireminecraft/customevents/EntityEvents.java
new file mode 100644
index 0000000..d4cb6bc
--- /dev/null
+++ b/src/main/java/com/empireminecraft/customevents/EntityEvents.java
@@ -0,0 +1,53 @@
+package com.empireminecraft.customevents;
+
+import org.bukkit.Chunk;
+import org.bukkit.entity.Entity;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+
+public class EntityEvents {
+    public static class EntitySpawnEvent extends Event {
+        private static final HandlerList handlers = new HandlerList();
+
+        private final Entity entity;
+
+        public EntitySpawnEvent(Entity entity) {
+            this.entity = entity;
+        }
+
+        public Entity getEntity() {
+            return entity;
+        }
+
+        @Override
+        public HandlerList getHandlers() {
+            return handlers;
+        }
+
+        public static HandlerList getHandlerList() {
+            return handlers;
+        }
+    }
+    public static class EntityDespawnEvent extends Event {
+        private static final HandlerList handlers = new HandlerList();
+
+        private final Entity entity;
+
+        public EntityDespawnEvent(Entity entity) {
+            this.entity = entity;
+        }
+
+        public Entity getEntity() {
+            return entity;
+        }
+
+        @Override
+        public HandlerList getHandlers() {
+            return handlers;
+        }
+
+        public static HandlerList getHandlerList() {
+            return handlers;
+        }
+    }
+}
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 3559bfc..cf523c7 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1140,6 +1140,7 @@ public abstract class World implements IBlockAccess {
         }
 
         entity.valid = true; // CraftBukkit
+        new com.empireminecraft.customevents.EntityEvents.EntitySpawnEvent(entity.bukkitEntity).callEvent(); // EMC
     }
 
     protected void b(Entity entity) {
@@ -1147,6 +1148,7 @@ public abstract class World implements IBlockAccess {
             ((IWorldAccess) this.u.get(i)).b(entity);
         }
 
+        new com.empireminecraft.customevents.EntityEvents.EntityDespawnEvent(entity.bukkitEntity).callEvent(); // EMC
         entity.valid = false; // CraftBukkit
     }
 
-- 
1.9.1

